### DISCUSSION TITLE
Need to use Token-2022

### DISCUSSION THREAD

**Discussion by pfleeter** [bmresearch/Solnet#444] [Show and tell]
Might be simple, but even though the TokenProgram.ProgramIdKey is static and readonly, you can still _set_ the KeyBytes on it in a static constructor to swap the library over to Token-2022, like this:

```
static Blockchain()
{
    TokenProgram.ProgramIdKey.KeyBytes = new PublicKey("TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb").KeyBytes;
}
```

Maybe others will find that helpful.

**Comment by pfleeter** [comment]
Well, this kind of worked.  I ran into problems because the "Key" member of ProgramIdKey doesn't update when you update KeyBytes, so the bytes are Token 2022 but the Key is the legacy program.  This causes problems in things like ATA creating where the Key is added to an account and then you get a cryptic "AccountLoadedTwice" error as a result.

My fix for that was to write a Instruction upgrader method which can find the legacy token key in the accounts list and swap them for the new token program.

This is all just hackery to work around lack of support without needing to fork things etc.

    public static TransactionInstruction UpgradeTokenProgramInInstruction(PublicKey upgradedTokenProgramIdKey, TransactionInstruction origInstruction)
    {
        // If there is no upgrade, do nothing
        if (upgradedTokenProgramIdKey == TokenProgram.ProgramIdKey)
        {
            return origInstruction;
        }
        
        var accountKeys = new List<AccountMeta>(origInstruction.Keys.Count);
        
        foreach (var k in origInstruction.Keys)
        {
            if (k.PublicKey == TokenProgram.ProgramIdKey)
            {
                accountKeys.Add(k.IsWritable
                    ? AccountMeta.Writable(upgradedTokenProgramIdKey, k.IsSigner)
                    : AccountMeta.ReadOnly(upgradedTokenProgramIdKey, k.IsSigner));
            }
            else
            {
                accountKeys.Add(k);
            }
        }
            
        var upgradedInstruction = new TransactionInstruction()
        {
            ProgramId = origInstruction.ProgramId,
            Keys = accountKeys,
            Data = origInstruction.Data
        };

        return upgradedInstruction;
    }

**Reply by jcwrx** [reply]
@pfleeter - I will give this a go, thanks.
